File: src/main/java/org/leetcode/bestTimeToBuyAndSellStock_121.java
package org.leetcode;

public class bestTimeToBuyAndSellStock_121 {
    public int maxProfit(int[] prices) {
        int low = prices[0];
        int res = 0;
        for(int curr: prices){
            if(curr < low){
                low = curr;
            }
            else{
                res = Math.max(res, curr - low);
            }
        }
        return res;
    }
}

File: src/main/python/BestTimeToBuyAndSellStock_121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        buy, sell = 0, 1
        while(sell < len(prices)):
            if prices[buy] < prices[sell]:
                res = max(res, prices[sell] - prices[buy])
            else:
                buy = sell
            sell += 1
        return res

    # DP
    def maxProfit(self, prices: List[int]) -> int:
        buy = float('inf')
        res = 0
        for p in prices:
            if p < buy:
                buy = p
            res = max(res, p - buy)
        return res